woocommerce/woocommerce-gutenberg-products-block
# Try dynamic blocks for rendering content (replacing shortcode output)

We've had some discussion about needing to extend the shortcodes (#554), the shortcodes causing issues with theme switches (#548), and potential future layout improvements (#371) won't be possible with the current shortcode rendering. This PR is an exploration to getting rid of the shortcode in a back-compat way.

This PR introduces a new abstract class `WGPB_Block_Grid_Base`, which handles building the query, fetching, and rendering products in the grid. Each block will extend this class and inject any specific query handling using , see [`WGPB_Block_Product_On_Sale` for an example.](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/245d0d9f27c0c158d0fbf612411f5ffb8aeae503/assets/php/class-wgpb-block-product-on-sale.php)

Also…

- Uses [block deprecation](https://wordpress.org/gutenberg/handbook/designers-developers/developers/block-api/block-deprecation/) to convert existing blocks to the new method once a post is edited.
- Changes some of the markup of the preview to match the markup we're now using on the frontend.

Adding this to other current grid-blocks should be this simple: https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/245d0d9f27c0c158d0fbf612411f5ffb8aeae503 (plus https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/55cb49fc2ebcee4ebb70c97a9b7f5da8e6549277)

#### Accessibility

- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/) – depends on the theme

### Screenshots

Storefront:

![Screen Shot 2019-05-10 at 6 32 03 PM](https://user-images.githubusercontent.com/541093/57560023-80fbaa00-7352-11e9-8a8f-deaef3c0985a.png)

Same post in Twenty Nineteen (there's a bug with the On Sale flag here)

![Screen Shot 2019-05-10 at 6 36 07 PM](https://user-images.githubusercontent.com/541093/57560035-86f18b00-7352-11e9-971c-25d3e95c3e63.png)

The styles are a little different here– we're not using an `h2` since semantically we can't really know if that's the correct heading level, instead we're listing everything as a `div`. CSS can be used to title-style `.wc-block-grid__product-title`.

### How to test the changes in this Pull Request:

The only blocks that have changed here are Newest and On Sale. If this looks good, we can merge it & update the rest of the grid-blocks.

1. Before checking out this branch, create a few posts with assorted blocks, Newest and On Sale (but also a few others)
2. Check out the branch, build
3. Load any of your existing posts in the editor, there should be no broken blocks
4. The posts should look correct on the frontend, and show the correct products/order/category/etc.
5. Create a new post with Newest Products and/or On Sale Products, it should work

Bonus points: try to reproduce #548 – this PR should also fix that issue.

cc @tiagonoronha Can you check this PR out, and see if it's better for theme styling?
